File: frontend/src/services/chatService.js
import api from './api';

// Chat Services
export const chatService = {
    // Create or get existing chat room
    createChatRoom: async (vendorId, shopkeeperId) => {
        const response = await api.post('/chat/room', null, {
            params: { vendorId, shopkeeperId }
        });
        return response.data;
    },

    // Get all chat rooms for a vendor
    getVendorRooms: async (vendorId) => {
        const response = await api.get(`/chat/active-rooms/vendor/${vendorId}`);
        return response.data;
    },

    // Get all chat rooms for a shopkeeper
    getShopkeeperRooms: async (shopkeeperId) => {
        const response = await api.get(`/chat/active-rooms/shopkeeper/${shopkeeperId}`);
        return response.data;
    },

    // Send a text message
    sendMessage: async (messageData) => {
        const response = await api.post('/chat/message', messageData);
        return response.data;
    },

    // Get all messages in a chat room
    getMessages: async (roomId) => {
        const response = await api.get(`/chat/room/${roomId}/messages`);
        return response.data;
    },

    // Send estimation (vendor creates order through chat)
    sendEstimation: async (chatRoomId, senderId, orderData) => {
        const response = await api.post('/chat/estimation', orderData, {
            params: { chatRoomId, senderId }
        });
        return response.data;
    },

    // Send invoice (vendor confirms order through chat)
    sendInvoice: async (chatRoomId, senderId, orderId, invoiceContent) => {
        const response = await api.post('/chat/invoice', invoiceContent, {
            params: { chatRoomId, senderId, orderId },
            headers: { 'Content-Type': 'text/plain' }
        });
        return response.data;
    }
};

export default chatService;

File: backend/Ecommerce B to B/Ecommerce B to B/Service/ChatService.cs
using Ecommerce_B_to_B.Data;
using Ecommerce_B_to_B.DTOs;
using Ecommerce_B_to_B.Models;
using Microsoft.EntityFrameworkCore;

public class ChatService : IChatService
{
    private readonly ApplicationDBContext _context;

    public ChatService(ApplicationDBContext context)
    {
        _context = context;
    }

    // TEXT MESSAGE
    public async Task<Message> SendTextMessageAsync(SendMessageDto dto)
    {
        var message = new Message
        {
            ChatRoomId = dto.ChatRoomId,
            SenderId = dto.SenderId,
            SenderRole = dto.SenderRole,
            MessageType = "Text",
            Content = dto.Content
        };

        _context.Messages.Add(message);
        await _context.SaveChangesAsync();

        return message;
    }

    // INVOICE / ESTIMATION - Chat-Driven Order System
    public async Task<Message> SendInvoiceAsync(SendInvoiceDto dto)
    {
        Order order;

        if (dto.MessageType == "Estimation")
        {
            // CREATE NEW ORDER when Estimation is sent
            order = new Order
            {
                VendorId = dto.VendorId,
                ShopkeeperId = dto.ShopkeeperId,
                TotalAmount = dto.TotalAmount,
                OrderStatus = "Estimated"
            };

            _context.Orders.Add(order);
            await _context.SaveChangesAsync();
        }
        else if (dto.MessageType == "Invoice")
        {
            // UPDATE EXISTING ORDER when Invoice is sent
            order = await _context.Orders
                .Where(o => o.VendorId == dto.VendorId && o.ShopkeeperId == dto.ShopkeeperId)
                .OrderByDescending(o => o.CreatedAt)
                .FirstOrDefaultAsync();

            if (order == null)
            {
                throw new InvalidOperationException("No estimation found. Send estimation first.");
            }

            // Update existing order
            order.TotalAmount = dto.TotalAmount;
            order.OrderStatus = "Pending";
            await _context.SaveChangesAsync();
        }
        else
        {
            throw new ArgumentException("MessageType must be 'Estimation' or 'Invoice'");
        }

        // Create chat message linked to order
        var message = new Message
        {
            ChatRoomId = dto.ChatRoomId,
            SenderId = dto.VendorId,
            SenderRole = "Vendor",
            MessageType = dto.MessageType,
            Content = dto.Content,
            OrderId = order.OrderId
        };

        _context.Messages.Add(message);
        await _context.SaveChangesAsync();

        return message;
    }

    // GET CHAT HISTORY
    public async Task<List<Message>> GetMessagesAsync(int chatRoomId)
    {
        return await _context.Messages
            .Where(m => m.ChatRoomId == chatRoomId)
            .OrderBy(m => m.SentAt)
            .ToListAsync();
    }
}
